### PR TITLE
Make node module, add errors, zerofilling & tests

### PR DESCRIPTION
**Pull request by dawaa**
Added unit tests to the core functionality and also expose the methods through the `prototype`-chain to allow using them throughout our tests in an isolated matter.

### CODE REVIEW

**Comment by dawaa** [comment]
This is BTW a continuation from this PR, https://github.com/Advisa/kontonummer.js/pull/1.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: Should all these really be public? I'd argue that at the very least `mod10` & `mod11` are private functions, if not all should be private. I think it's one thing if this was a class that you initiated with an account number and could run method on like `getAccountNumber` or `validate`, but as this is currently constructed, `kontonummer.js` is a pure matching and validation function.

Sure, they need to be exposed somehow to test them, but I don't think they should be marked as an open API, rather marked as private both with a comment and a `_` prefix.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: Why expose a private variable in the public prototype? Isn't it better to keep it private within the scope?

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: How about being consistent with the project and use semicolons in this whole file?

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: Why is function not being used above in `validateChecksum` where the same thing is assigned to `ctrlNum`? Considering there are tests for this function, it seems resonable to actually use it internally.

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: Isn't this message a bit misleading? It returns part of the bank account number based on `mockBank.lengths.control`, not `mockBank.modulo` as the description may suggest - it doesn't filter banks based on modulo and doesn't use that value in any way.

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: What's the reason for testing multiple account numbers here that match the same bank? The only reason I see for testing multiple account numbers, would be if they were of different lengths.

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: How about keeping minimal mocks, as in just what the function needs to better describe what the tests is doing? To me this test somehow suggests that the regex is relevant to this test, because the `fakeAccountNumbers` match it, while in fact it's irrelevant - the only thing relevant here is the `lengths.clearing`.

Or maybe even better - why does the function need to accept the whole bank object? Why not just the `lengths.clearing` as an argument?

**Review by alexteg-advisa — COMMENTED** [review]
Don't forget to update the minified file or remove it.

File: kontonummer.js
Comment (dawaa) [reply kontonummer.js]: This was an idea I had early on but forgot to remove it once I changed mindset

**Comment by dawaa** [comment]
@alexteg-advisa I have adressed your feedback

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: Isn't this comment superfluous? It contains the exact same words as the function name that is repeated several times...

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: This test name is the same as the previous one. How about differentiating it? Like `returns a 10/11 digit control number`. But I would also question if there is any purpose to having two tests that do the same thing, just with different length, effectively just testing the `.substr()` function.

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: I'm still wondering:

> What's the reason for testing multiple account numbers here that match the same bank?

https://github.com/Advisa/kontonummer.js/pull/2#discussion_r204045062

**Review by alexteg-advisa — COMMENTED** [review]
Nicely cleaned up.

File: kontonummer.js
Comment (dawaa) [reply kontonummer.js]: This was an inside joke with myself.. yeah I need to get out more. I'll have it removed in a fixup

File: tests/kontonummer.js
Comment (dawaa) [reply tests/kontonummer.js]: You're right they could actually be one test

File: tests/kontonummer.js
Comment (dawaa) [reply tests/kontonummer.js]: I'm not entirely sure what my thought was, but I think I wanted to be more verbose on what the input/output are to demonstrate more than just one example

**Comment by dawaa** [comment]
@alexteg-advisa I've adressed your feedback.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: I can't comment on the right line, but regarding `snake_case` here, do we want to be consistent with `isValid` and `matchedBanks` returned in the constructor on line `400` as well that are camelCased now? Seems a bit weird to return `camelCase` properties with an array of objects with `snake_case` properties.

Don't forget to change the `README` file as well.

File: tests/kontonummer.js
Comment (alexteg-advisa) [inline tests/kontonummer.js]: Sorry if I was unclear. I didn't mean that both asserts need to be in a single test. I think it's good to only have one assert per test to easily see what is wrong if it breaks. However, my point was that by having multiple asserts testing the same function that only uses `substr` internally, basically means you are testing the `substr` method rather than the specifics of _how_ the function is using `substr`.

What is unique about this function, is what part of the string it takes out (X characters from the end). The first assert checks that by ensuring that the first number `1` is not included in the expected return value of `_getCtrlNum()`. However, checking if it picks 11 of 11 characters or 10 of 11 characters is kind of testing the same thing.

If you want to check edge cases for regression testing in case we desire to change the internal implementation, such as because of the latest recommendation in the [9th edition of ECMA-262](https://www.ecma-international.org/ecma-262/9.0/index.html#sec-string.prototype.substr) of not using [`substr`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) in new code ([`slice`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/slice) could replace it), maybe checking the result of picking 12 characters of 11 would be a more useful test? That could potentially return different things depending on the internal implementation and could happen for real, because of the changes Marcin made in a04b62af27e2df201279609b5a541c986010fa7b that mean the wrong digits may be included in the mod10/11 check, which may produce unexpected results. For example: If a bank account number needs to be 11 characters, the user enters 10 missing the last digit, the mod10 check may still pass, but on the wrong digits.

**Review by alexteg-advisa — CHANGES_REQUESTED** [review]
While all tests pass, the library as a whole is not working at all. You may want to add some test for the `kontonummer` constructor. Currently it will always return the error `unknown_clearing_number`, because no bank is matched, since https://github.com/Advisa/kontonummer.js/blob/c590c446c3bc8dde12b44ddcc03ea125cb0ff94e/kontonummer.js#L384 needs to use snake_case `bank_name` that `validateChecksum` returns instead of `bankName`.

**Comment by dawaa** [comment]
@alexteg-advisa I've adressed your feedback, though I wasn't entirely sure if I had covered the `_getCtrlNum()` test as you explained. If I didn't, please talk to me directly so I can fully grasp your thoughts

**Comment by dawaa** [comment]
@alexteg-advisa almost forgot to ignore one file, I'm on it

**Comment by dawaa** [comment]
@alexteg-advisa I've updated the minified file, ignored another file and added the `engines` node.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: Isn't it safer to either do `typeof accountNumber === 'number'` or `accountNumber instanceof Number`? If you do  `accountNumber.constructor === Number` that could be circumvented:
```js
const foobar = {};
foobar.constructor = Number;
foobar.constructor === Number // is true
```
also `(null).constructor` will throw an error, but that you have guarded against with `!accountNumber`.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: This is kind of superfluous. The next check (`!accountNumber`) will catch any falsy values (`null`, `undefined`, `false`, `''`) that this catches.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: This will prevent many users from entering valid account numbers. Many banks provide account numbers in formats like `9159-123.456-7` or `8105-9,876 543 210-9` containing everything from spaces, dashes, dots, commas and other characters that are just used for visual formatting. However, we want to allow customers to just copy-paste such account numbers. We should only care about digits. This is also superflous, since we only pick out digits a few lines down `accountNumber.replace(/\D/g, '')`.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: You may want to cast the account number to string at this point `String(accountNumber).replace(/\D/g, '')`. In spite of the checks above, one may still pass in an object that has a `toString` method (with the above checks it is cast to string when for example doing `!/^\d+$/.test(accountNumber)`) and that will therefore pass the above, but fail here, since it attempts to call `replace()` on an object that may not have that method.

### Example:

```js
accountNumber = ()=> {}
accountNumber.toString = ()=> '123'
!accountNumber // is false
!/^\d+$/.test(accountNumber) // is false
accountNumber.constructor === Number // is false
typeof accountNumber === 'number' // is false
accountNumber instanceof Number  // is false
```

alternatively, we could also decide to only accept strings and check that above instead of the "not number" check with either:

```js
typeof accountNumber !== 'string'  // is true
!(accountNumber instanceof String)  // is true
```

**Review by alexteg-advisa — CHANGES_REQUESTED** [review]
Great, considerable improvement. Only one real blocking issue regarding formatting characters in account numbers, otherwise just 3 suggestions.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: When I think about it, do we really want to throw an error in a validator? Isn't it more reasonable to just return the account number as invalid with a generic error code?

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: If you check for string above, this is no longer needed. If you want to keep this, then it needs to return a consistent invalid object with `errors` that contains some error code and an empty array of `matched_banks` like below https://github.com/Advisa/kontonummer.js/blob/22481925b0b5eaeee3b37484e9ac3fd8e6dc2f63/kontonummer.js#L409-L413

Or remove `matched_banks` in other cases and document it in the `README`. There needs to be a consistent API that can be relied upon regarding the expected response format.

File: kontonummer.js
Comment (alexteg-advisa) [inline kontonummer.js]: I'm realizing this may become a bit misleading with the error messages. If you enter the correct clearing number `3300`, but an account number that doesn't match the rest of the regex, you still get the error `unknown_clearing_number`, which may be confusing. I think we should just leave this to check the clearing number only.

If we want specific validators for the account number, those should be added and checked separately with a different error message, but we can leave that for the future. For now, the length check should at least ensure the length is correct.

File: kontonummer.js
Comment (dawaa) [reply kontonummer.js]: ### typeof

`typeof` really depends on what we're expecting, in this case it could work. Though it's good at differentiating between Primitives it's not very good at diff'ing different objects.

`typeof`
```js
typeof 3 // "number"
typeof "abc" // "string"
typeof {} // "object"
typeof true // "boolean"
typeof undefined // "undefined"
typeof function() {} // "function"

// And where it falls short..
typeof [] // "object"
typeof null // "object"
typeof new Date() // "object"
typeof new Regex() // "object"
```

### instanceof

Meanwhile `instanceof` tells you whether an object is an instance of a certain type. In this case the `type` here is the constructor, still.

```js
function Animal(){}
var a = new Animal()
a instanceof Animal // true

// And checking against the property `constructor`
a.constructor === Animal // true
```

Now the difference here is that the `.constructor`-type check doesn't walk up the `prototype` chain (which `instanceof` does and performance wise is worse).

Also running the type check using the `constructor` property will throw `TypeError` if the object you're checking is either `null` or `undefined`. Which could be nice to have at some cases where you want to make sure what you get is right otherwise exit out entirely.

Now looking at how `instanceof` deals with certain types.

`instanceof`

```js
[1, 2, 3] instanceof Array // true
/abc/ instanceof RegExp // true
({}) instanceof Object // true
(function() {}) instanceof Function // true

// but then..
3 instanceof Number // false
true instanceof Boolean // false
'abc' instanceof String // false

// "null" or "undefined" whcih are not an instance of anything
null instanceof Boolean // false
undefined instanceof Array // false
```

Here we see that `instanceof` doesn't work well with Primitives.

### constructor type checking

Now what if we use the `constructor´ property to check the same types.

`constructor`

```js
(3).constructor === Number // true
true.constructor === Boolean // true
'abc'.constructor === String // true
```

They result is now correct, the magic here is what's going on behind the scenes. Accessing the property of a Primitive object javascript will wrap it with an Object wrapper leaving us with

```js
// The wrapping is nothing we see, it's done behind the scenes
new Number(3) instanceof Number // true
new Boolean(true) instanceof Boolean // true
new String('abc') instanceof String // true
```

So in the end I guess it depends on what we want, but I'm more leaning towards the `constructor` method as the result is more precise, though it has its pitfalls such as being able to check against the parent of a deriving class, the `TypeError` it will throw if you try accessing it on `null` and `undefined` and lastly it will not walk up the `prototype` chain.

**Comment by dawaa** [comment]
@alexteg-advisa I've addressed your issues once more, please have a look.

File: kontonummer.js
Comment (alexteg-advisa) [reply kontonummer.js]: I'm aware of the `typeof` having shortcomings in regards to objects, but for checking primitive types like number or string that want here, it should be fine. I'm not sure what I was thinking about regarding `instanceof`, I agree it's probably the worst of the three in this case.

As long as we keep the `!accountNumber` check first, using `accountNumber.constructor` will be fine also with `null` and `undefined`.

**Comment by dawaa** [comment]
@alexteg-advisa I had forgotten to update the `readme.md` file, I did that in the `fixup` comit 2fa95fe. Also, I will update the minified script once this is approved.

**Review by alexteg-advisa — COMMENTED** [review]
Code and functionality look good!

However, we need to clean up the commit structure. It's become a bit messy with two pull requests. Considering you have fixup commits here for #1, there are two ways to solve this:

* Move your commits to #1 branch `marcin/improvements` and push them to that pull request and do fixups and updating the minified file there and close this PR #2.
* Close #1 and change the base of this PR to `master` and do fixups and updating the minified file here.

in any case, I suggest a final review round once everything is squashed and ready for merge.

**Review by alexteg-advisa — APPROVED** [review]
I've squashed the fixup commits and cleaned up the git history, added some info in git commit messages, upped the version and added a release description to the README.

@dawaa If you think my git changes are alright, feel free to merge.

File: kontonummer.js
Comment (dawaa) [inline kontonummer.js]: We're a comma short here. Also remove the trailing comma on the line below this one to follow the same format throughout the file.

**Comment by alexteg-advisa** [comment]
Changed version change to major, since it's a breaking API change in what `kontonummer` returns.

File: README.md
Comment (dawaa) [inline README.md]: `node` with a capital N? `deljerade`, did you mean `detaljerade`?

**Review by dawaa — COMMENTED** [review]
There is a typo in the commit message of 4caacae.

 `This assits users by automatically ...`

**Comment by alexteg-advisa** [comment]
@dawaa Fixed all

**Comment by dawaa** [comment]
well done, merging 🥇

**Pull request merged**
